src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -62,114 +64,154 @@ void memory_snapshot_harness_generatort::handle_option(
   {
+    const auto &goto_program = goto_function->second.body;
     const auto opt_it = goto_program.get_target(*location_number);
 
     if(!opt_it.has_value())
     {
       throw invalid_command_line_argument_exceptiont(
         "no instruction with location number " +
-          std::to_string(*location_number) + " in function " +
-          id2string(function),
+        std::to_string(*location_number) + " in function " +
+        id2string(entry_function_name),
         "--initial-location");
     }
+  }
 
-    it = *opt_it;
+  if(goto_functions.function_map.count(INITIALIZE_FUNCTION) == 0)
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "invalid input program: " + std::string(INITIALIZE_FUNCTION) +
+        " not found",
+      "<in>");
   }
-  else
+
+  const symbol_tablet &symbol_table = goto_model.symbol_table;
+  const symbolt *called_function_symbol = symbol_table.lookup(entry_function_name);
+
+  if(called_function_symbol == nullptr)
   {
-    it = start_it;
+    throw invalid_command_line_argument_exceptiont(
+      "function `" + id2string(entry_function_name) + "` not found in the symbol table",
+      "--initial-location");
   }
 
-  auto ins_it1 = goto_program.insert_before(start_it);
-  ins_it1->make_goto(goto_program.const_cast_target(start_it));
-  ins_it1->guard = var;
+  if(symbol_table.has_symbol(harness_function_name))
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "harness function `" + id2string(harness_function_name) +
+      "` already in "
+      "the symbol table",
+      "--" GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT);
+  }
+}
+
+void memory_snapshot_harness_generatort::add_init_section(
+  goto_modelt &goto_model) const
+{
+  goto_functionst &goto_functions = goto_model.goto_functions;
+  symbol_tablet &symbol_table = goto_model.symbol_table;
+
+  goto_functiont &goto_function = goto_functions.function_map[entry_function_name];
+  const symbolt &function_symbol = symbol_table.lookup_ref(entry_function_name);
+
+  goto_programt &goto_program = goto_function.body;
+
+  // introduce a symbol for a Boolean variable to indicate the point at which
+  // the function initialisation is completed
+  symbolt &func_init_done_symbol = get_fresh_aux_symbol(
+    bool_typet(),
+    id2string(entry_function_name),
+    "func_init_done",
+    function_symbol.location,
+    function_symbol.mode,
+    symbol_table);
+  func_init_done_symbol.is_static_lifetime = true;
+  func_init_done_symbol.value = false_exprt();
 
-  auto ins_it2 = goto_program.insert_after(ins_it1);
-  ins_it2->make_assignment(code_assignt(var, true_exprt()));
+  const symbol_exprt func_init_done_var = func_init_done_symbol.symbol_expr();
 
-  auto ins_it3 = goto_program.insert_after(ins_it2);
-  ins_it3->make_goto(goto_program.const_cast_target(it));
+  // initialise func_init_done_var in __CPROVER_initialize if it is present
+  // so that it's FALSE value is visible before the harnessed function is called
```

Comment:
That still doesn't make a lot of sense to me - you should be able to initialise it in the harness function without modifying `__CPROVER_initialize` (or indeed assuming there even _is_ a `__CPROVER_initialize`)?